### DISCUSSION TITLE
STM32F4-SKR-Pro-V1_2 Compile Error

### DISCUSSION THREAD

**Discussion by landtree** [Paciente8159/uCNC#719] [Config and issues] [answered]
Hello All,

I have been banging my head at this for awhile now.

I am using VsCode/PlatformIO, and used the web configurator to set up my files.  Change my environment to STM32F4-SKR-Pro-V1_2

I am using TMC drivers and the folder is in the Modules folder. If that makes any difference.

When I go to compile, I get this error:
compiling .pio\build\STM32F4-SKR-Pro-V1_2\SrcWrapper\src\HAL\stm32yyxx_hal.c.o
Compiling .pio\build\STM32F4-SKR-Pro-V1_2\SrcWrapper\src\HAL\stm32yyxx_hal_adc.c.o
Compiling .pio\build\STM32F4-SKR-Pro-V1_2\SrcWrapper\src\HAL\stm32yyxx_hal_adc_ex.c.o
Compiling .pio\build\STM32F4-SKR-Pro-V1_2\SrcWrapper\src\HAL\stm32yyxx_hal_can.c.o
Compiling .pio\build\STM32F4-SKR-Pro-V1_2\SrcWrapper\src\HAL\stm32yyxx_hal_cec.c.o
Compiling .pio\build\STM32F4-SKR-Pro-V1_2\SrcWrapper\src\HAL\stm32yyxx_hal_comp.c.o
<command-line>: fatal error: variant_BLACK_F407ZX.h: No such file or directory

I am not sure what to change at this point.

**Comment by Paciente8159** [comment]
Hi there. I'm AFK. As soon as I get the chance to test I'll give some feedback. Did you configured manually or used the web builder.
If you used the web builder please share the config files

**Reply by landtree** [reply]
Take your time! Thank you for the help. Here are my files from the web builder:

[ucnc_build.json](https://github.com/user-attachments/files/16474373/ucnc_build.json)
[boardmap_reset.txt](https://github.com/user-attachments/files/16474363/boardmap_reset.txt)
[boardmap_overrides.txt](https://github.com/user-attachments/files/16474364/boardmap_overrides.txt)
[cnc_hal_reset.txt](https://github.com/user-attachments/files/16474365/cnc_hal_reset.txt)
[cnc_hal_overrides.txt](https://github.com/user-attachments/files/16474366/cnc_hal_overrides.txt)

**Reply by landtree** [reply]
Follow up!

I ended up nuking my platformio and forced it to rebuild my platform files. I now have a new problem :)

Thank you!

![image](https://github.com/user-attachments/assets/8651b7cf-220c-4e52-93ae-f92fd0353654)

**Comment by Paciente8159** [comment]
Mine compiles fine.
![image](https://github.com/user-attachments/assets/5a8f9bed-9b61-4899-a015-4f73c9d2606d)

Maybe you have a broken cached files or broken framework.
Open a PlatformIO: New Terminal an type this command
`pio pkg list -eSTM32F4-SKR-Pro-V1_2`

This is my output:
![image](https://github.com/user-attachments/assets/940a45b0-87f4-40fc-a1e9-6b2d60f5d3c6)

**Reply by landtree** [reply]
Yeah, that seemed to have done it. I can compile the bin now.

New problem though, I seem to not be able to talk to the board via USB. 

At least I have some forward momentum! Thank you!

**Reply by Paciente8159** [reply]
does the board get recognized as a serial port on your PC?
Or you get an error saying that the driver was not recognized?

**Reply by landtree** [reply]
It does not show up as a device on my PC.

If I install marlin on it, it does show up.

**Reply by Paciente8159** [reply]
hummm. Ok I will investigate

**Reply by Paciente8159** [reply]
I don't have that board so if you are willing to help me out solving this it would be great.
The board map has a typo. Can you modify the DOUT32_BIT and DOUT32_PORT to DOUT31 in the skr boardmap file please (since port DOUT32 does not exist).
Then reflash your board.
You should get a yellow led blinking every second.
If not that is an indicator that the firmware is not running at all.

Thanks

**Reply by landtree** [reply]
I modified the bit and port.
![image](https://github.com/user-attachments/assets/e546ea01-b469-467b-9c51-0f86879d9f01)

Flashed the bin and I am not getting any blinking light.

**Reply by Paciente8159** [reply]
Thanks. Ok. Then that is a clue. The firmware is not running. Tomorrow I will try to investigate if there is a bad clock configuration. Are you uploading the firmware via sd card or via stlink?

**Reply by landtree** [reply]
I am uploading via a new SD card.

**Reply by Paciente8159** [reply]
Actually that can also be a misconfiguration.
You modified the config files. But the boardmap reset file deletes that. Make sure that that configuration is on the boardmap_override file too.

**Reply by Paciente8159** [reply]
I've patched master branch with #723
If you could retry with this it would be great. This would be important to determine if it's a problem of the firmware not running or simply  a problem with TinyUSB library.

**Reply by landtree** [reply]
I am away from the board until Monday, but I will do it first thing and follow up with an update. I appreciate all the help!

**Reply by Paciente8159** [reply]
Thanks. I'm waiting to get a MKS Robin Nano v3.1 that is somewhat similar to that one.
I hope I'm able to get my hands on it by the end of next week.

**Reply by landtree** [reply]
Alright, I am back with an update. I pulled the master branch with #723. I reran the configurator using the generic prebuilt settings for the SKR Pro 1.2 and modified nothing. My computer still does not see a USB port when plugged in. 

I then tried changing the UART Port to 1. Still no connection. In both attempts, I did not have any activity light; it blinks rapidly only on the boot(I assume this is the upload indicator), then nothing. 

I feel like I must be setting something in the configuration incorrectly.

I also get this warning in my VS code:
![image](https://github.com/user-attachments/assets/68ce31e0-caf3-468b-a619-8fb431e3b4cd)
 
Not sure if it is relevant...

**Reply by Paciente8159** [reply]
First thanks for you feedback.
That confirms the firmware is not running properly.
While I wait for the MKS board to arrive I'll try to peek at Marlin's configuration for that board to see if I missed something.

I'll give you feedback as soon as I find anything relevant.

That warning is not relevant.

**Reply by landtree** [reply]
I'm not sure if this will help you find out the issue, but I also tried using grblHAL and used the generic preset builder for that with no modifications. I was able to see and connect over USB. 

Thanks again for helping me!

**Reply by Paciente8159** [reply]
Thanks for the feedback. I'll take a peek at that too.

**Reply by Paciente8159** [reply]
Ok this might not be the problem but I suspect the base board that serves as a template for the SKR Pro v1.2 might be running a cristal witha different frequency and that might cause the firmware to simply not run at all.
Tomorrow I will investigate deeper and give you some feedback.

Thanks

**Reply by Paciente8159** [reply]
You can try to setup the cristal frequency by adding the option [at the end of this line](https://github.com/Paciente8159/uCNC/blob/3f1a64e53efe606159199383fd938d7078713740/uCNC/src/hal/boards/stm32/stm32.ini#L61) by adding this:

` -D HSE_VALUE=8000000`

But I will confirm this be looking at the stm32arduino core tomorrow.

**Reply by landtree** [reply]
I am still using the default settings from the web builder and updated that line with the HSE_VALUE. It compiled just fine with no issues, but I still don't get a blinking light, nor do I have a USB connection.  Thanks again for all the effort!

**Reply by Paciente8159** [reply]
Again thanks for trying this. Another culprit for this kind of problem would be an incorrect offset for the vector table.
I tried to compile adding these  option `-DHSE_VALUE=8000000 -DVECT_TAB_BASE_ADDRESS=0x08000000 -DVECT_TAB_OFFSET=0x8000`

This got me a bunch of `warning: "VECT_TAB_OFFSET" redefined`
I assumed the PIO configuration [board_build.offset  = 0x8000](https://github.com/Paciente8159/uCNC/blob/3f1a64e53efe606159199383fd938d7078713740/uCNC/src/hal/boards/stm32/stm32.ini#L58C1-L58C29) would take care of this.

Out of curiosity I built the project enabling verbose and I got this:

`arm-none-eabi-gcc -o .pio\build\STM32F4-SKR-Pro-V1_2\FrameworkArduinoVariant\generic_clock.c.o -c -std=gnu99 -std=gnu17 -Wall -fdata-sections -ffunction-sections -fno-exceptions -mcpu=cortex-m4 -mthumb -mfpu=fpv4-sp-d16 -mfloat-abi=hard -Os -ffunction-sections -fdata-sections -nostdlib --param max-inline-insns-single=500 -DPLATFORMIO=60115 -DSTM32F4 -DSTM32F407xx -DSTM32F40_41xxx -DHAL_TIM_MODULE_DISABLED -DHAL_EXTI_MODULE_DISABLED -DHAL_UART_MODULE_ONLY -DBOARD=BOARD_SKR_PRO_V1_2 -DHSE_VALUE=8000000 -DVECT_TAB_BASE_ADDRESS=0x08000000 -DVECT_TAB_OFFSET=0x8000 -DSTM32F4xx -DARDUINO=10808 -DARDUINO_ARCH_STM32 -DNDEBUG -DARDUINO_BLACK_F407ZG -DBOARD_NAME=\"BLACK_F407ZG\" -DHAL_UART_MODULE_ENABLED -DUSE_FULL_LL_DRIVER -DVARIANT_H=\"variant_BLACK_F407ZX.h\" -DCUSTOM_PERIPHERAL_PINS -DVECT_TAB_OFFSET=0x0 ....(continues)`

At the very end of this `VECT_TAB_OFFSET `is being redefined to offset 0. This means that on boot the main application is most probably just running the bootloader again instead of jumping to the offset address and running the main.
I will see what alternatives I have to go around this. I believe the firmware is being written in the correct offset. Just the vector table is being pointed at the wrong place.

**Reply by Paciente8159** [reply]
After some digging I discovered the issue with the vector table.

It seems Marlin (from which I based my configurations is locked to version 12.1 I believe. So this was the way to set the offset:
`board_build.offset = 0x8000`

I did not locked the framework version so it uses the latest stable (currently 17.5). This was modified in December/2023 to this:
`board_build.flash_offset = 0x8000`

this is again one warning bell that I need to lock framework versions to prevent these type of issues.

If you want to give a last try at it, please modify the SKR Pro PIO settings to this

```
[env:STM32F4-SKR-Pro-V1_2]
extends = common_stm32
board = black_f407zg
board_build.flash_offset = 0x8000
board_build.f_cpu = 168000000L
board_upload.offset_address = 0x08008000
board_upload.maximum_ram_size = 196608
board_upload.maximum_size = 1048576
build_flags = ${common_stm32.build_flags} -D BOARD=BOARD_SKR_PRO_V1_2 -D HSE_VALUE=8000000
```

**Reply by landtree** [reply]
Sadly, there is still nothing. Here is what I did: I fully cleaned, uninstalled, and reinstalled the libraries, just in case. I am still using the default config from the webbuilder for the SKR 1.2. I added the above code to the stm32.ini and replaced the old code.

I manually ran the build to get the verbose output as you might see something useful in that.
[OUTPUT.txt](https://github.com/user-attachments/files/16550784/OUTPUT.txt)

**Reply by Paciente8159** [reply]
Oh man I was really hoping this was the issue.
Apart from your output being unusually small (I assume you omitted all the middle stuff) I can see that the vector table is now correct

`.isr_vector           392   134250496`  => this is address 0x08008000 the correct position

I guess I really have to check how this goes when I get the MKS board in my hands and try to flash it.
Again many thanks for all your tests.

**Reply by landtree** [reply]
Here is the full output incase you want to see it. 
[OUTPUT.txt](https://github.com/user-attachments/files/16551167/OUTPUT.txt)

**Comment by Paciente8159** [comment]
I've created #726 to address this issue.
As soon as I can confirm this solves the issue I'll push it.
Again thanks for reporting this and the tests.

**Reply by Paciente8159** [reply]
I've got my board.
![20240809_135650](https://github.com/user-attachments/assets/18268059-af02-424f-87bf-77f95fc00f51)
Don't know if I'll be able to play with it this weekend though.
I'll give some feedback as soon as I have news.

**Reply by Paciente8159** [reply]
@landtree I've added patches #729 and #726 and I hope this will solve the issue for the SKR board too.

Like you, after loading the firmware the board was not giving any life signs and the USB port was not showing. Also I was not being able to debug the process.
I loaded Marlin and all worked like you said.

So I tried to clean the whole flash and just install the firmware without bootloader and no offsets. This allowed me to debug the code and the startup process. While debugging I noticed that the loader startup file never calls the `SystemClockUpdate` function to properly setup the clocks sources of the MCU (including the `HSE_VALUE` setting).
Adding that made µCNC work and communicate.

After cleaning the flash and reinstalling the bootloader I placed a copy of the firmware with the expected flash offsets and the firmware was updated and µCNC started running as expected.

So I hope this also solves the issue on your board.

**Reply by Paciente8159** [reply]
Just another note.
I've added an option (`RAM_ONLY_SETTINGS`) to prevent using the MCU flash to read and write the board settings (Grbl style settings).
This will prevent flash ware but will force you to send the configurations every time you powerup the board.

I will try in these next days make an addon module to allow using external I2C EEPROM available in the board.

**Reply by landtree** [reply]
Thanks, I will give this a try soon, in like a day and will report back!

**Reply by landtree** [reply]
Dang, no good! 

Cloned the Master and redownloaded generic SKR pro v1.2 config files.
Built with env:env:STM32F4-SKR-Pro-V1_2 and copied the bin over to sd card. Started it and no blink or connectivity, but my bin changes to CUR

Do you think its worth buying an STLink to see if that makes any difference?

**Reply by Paciente8159** [reply]
Oh man. I was really hopping this time it would work.

Again thanks for testing and feedback.
I'm not sure what is your experienced around C code and debugging with STM32 so probably no. If little then the STLink would not serve you much.
It would allow you to try the firmware directly at boot address (but this removes the bootloader) and you would have to reflash the bootloader later on to restore it, but I don't want you to risk your board over it.

The only thing that occurs to me is trying to get an SKR board myself...but I was hopping not needing to get another board when the MKS Robin is so similar.

Let me recheck the code and think for a couple of days on possible tests to do on your board as it is.

Can you confirm the chip model? It's and STM32F407ZG?

**Reply by Paciente8159** [reply]
By the way the file name change to cur means that it was renamed by the bootloader after writing the firmware..

**Reply by landtree** [reply]
Chip reads STM32F407ZGT6

![PXL_20240812_220241828](https://github.com/user-attachments/assets/37004548-9c51-4347-942e-f411cc1633af)

**Reply by Paciente8159** [reply]
@landtree

I made a modification that is in this [custom-stm32f4-clocks branch](https://github.com/Paciente8159/uCNC/tree/custom-stm32f4-clocks)
This basically runs a custom clock initialization that resets all clocks and vector table and reconfigures it again.
I tested this on the MKS Robin and worked.

Please try it and let me know if it shows any sign of life or not.

**Reply by landtree** [reply]
Sadly, no blinking yet.

I downloaded that repo and worked just out of that folder. Download a fresh config file again just to be consistent. No errors in the build, just the same as before.

I even flashed it twice just for the thrill of it.

**Reply by Paciente8159** [reply]
This is putting up quite the fight.
Can you try to upload it without the config files? Just a clean vanilla install without modules.

**Reply by Paciente8159** [reply]
Better yet. Try this binary.
[firmware.zip](https://github.com/user-attachments/files/16605670/firmware.zip)

**Reply by landtree** [reply]
I will send you an update in the morning!

**Reply by landtree** [reply]
I seem to have to continue to be the harbinger of bad news. With that firmware file I still don't have anything.

**Reply by landtree** [reply, marked as answer]
I don't know if this is useful, but I borrowed an ST-Link v2 from my friend. 

I haven't done anything other than connect to it via the stlink utility and sit and stare at it.

![image](https://github.com/user-attachments/assets/a7ef2ee8-58e4-4a26-827e-ed5f1ded24e3)

**Reply by Paciente8159** [reply]
Can you scroll down to see a bit above address 0x08008000?
A couple of lines beforw that address should also be visible.

Thanks

**Reply by landtree** [reply]
Let me know if you want any other screenshots.
![image](https://github.com/user-attachments/assets/7c029de3-eaa7-44a5-b1ea-a4f7d56eeca7)

**Reply by Paciente8159** [reply]
Awesome. Can you show me a 4 or 5 line above address 0x08008000?

**EDIT**
I can tell right now that signature does not look correct.

![image](https://github.com/user-attachments/assets/a032bec6-2fd0-40de-8e4f-e3467b122495)

**Reply by landtree** [reply]
![image](https://github.com/user-attachments/assets/41e0a142-07fc-475e-b584-0e3e6ec341c3)

**Reply by Paciente8159** [reply]
Here is how it looks on my MKS Robin.

See how the flash before is not written (all 0xFF)? That is the bootloader free space boundary.
The bootloader starts and if does not find any firmware to write jumps to the bootloader offset were it should find a 0x2000xxxx address.

Your bootloader seems to go beyond that. Actually beyond address 0x0800C654.


![image](https://github.com/user-attachments/assets/89b4da4e-99c9-4228-8316-e23d283bc027)

**Reply by Paciente8159** [reply]
Try this.
Click the disconnect target.
Write this address in the address input field 0x0800C600
Write this in the size input field 0xC000.
Press connect to target.

Scroll down and search were is the first 0x2000xxxx value after position 0x0800C654.

**Reply by landtree** [reply]
This is the only space with data written.

![image](https://github.com/user-attachments/assets/6177e706-3f26-4f8e-aeae-bfda224feb5f)

**Reply by Paciente8159** [reply]
No more data down?
Something is off.

**Reply by landtree** [reply]
0xFF all the from 0x0800C660 onward

**Reply by Paciente8159** [reply]
SKR does not have a repository of the bootloader. I only found this.
https://github.com/P-C-R/SKR_PRO_BOOTLOADER_REPAIR/blob/master/SKR_Pro_with_Bootloader.bin

I'll download and check.

**Reply by Paciente8159** [reply]
This is what it should look like.
Do you use a Marlin compilation you did or one of the supplied compilations at the SKR Pro github page?

![image](https://github.com/user-attachments/assets/65c4c599-6300-4779-9f1c-89eb3d761142)

**Reply by Paciente8159** [reply]
Wait a second.
Now that I look at your first print. That is the exact same signature as µCNC firmware.

Are you loading the firmware via SD card?
It's being written to address 0x08000000 if I'm seeing correct.

**Reply by landtree** [reply]
Yes, using the SD Card.

**Reply by Paciente8159** [reply]
Try to compile like this then, and then upload the firmware via SD Card.
This comments the offsets of the bootloader just to see if my assumption was correct.

![image](https://github.com/user-attachments/assets/56d819f9-b7d3-453f-bbc3-147d63079fef)

**Reply by landtree** [reply]
Okay.

![image](https://github.com/user-attachments/assets/8b5d5111-2d6d-416a-9736-72522912a130)


and then this is the x0800000

![image](https://github.com/user-attachments/assets/8fd1498e-f5b7-46e3-963e-93b65c221459)


![image](https://github.com/user-attachments/assets/c1e936db-26fc-43c8-87fb-70bce83d48cf)


Let me know if you want anything specific

**Reply by Paciente8159** [reply]
Did you compiled the firmware (just the vanilla version without modules) with the commented offsets and uploaded it via sd card to the board?
Still no communication?

**Reply by landtree** [reply]
I cloned the master and commented out the code, but I went to try it again, and this time.

The firmware did not used. It still has FIRMWARE.CUR & firmware.bin..
 🥹

**Reply by Paciente8159** [reply]
Are you able still upload Marlin?
You bootloader signature looks very odd.

**Reply by landtree** [reply]
I just copied over my bin that worked for the marlin build I have. 

It does not seem to want to copy that, either... It was uploading via the SD card ....

**Reply by Paciente8159** [reply]
Then probably the bootloader got corrupted.
Good thing you have the STLink.
Check this video explaining how to flash it.

https://www.youtube.com/watch?v=7O_ZOdnM0SU
and the bootloader
https://github.com/P-C-R/SKR_PRO_BOOTLOADER_REPAIR/blob/master/SKR_Pro_with_Bootloader.bin

**Reply by Paciente8159** [reply]
Were you able to re-flash it? Let me know if you are having any problem.

**Reply by landtree** [reply]
Alright, I just finished. With just that bootloader bin, I am able to see and connect.

I was going to compile a vanilla bin from the Master, unless you think I should test something specific?

**Reply by Paciente8159** [reply]
Yes now test with the vanilla master version. Don't change anything.

**Reply by landtree** [reply]
Alright, Vanilla untouched bin. Uploaded via SDcard.

I do not get a USB connection or blinking LED.

**Reply by Paciente8159** [reply]
Can you reconnect the STLink and show me the reading 5 or 6 lines above of address 0x08008000, please?

**Reply by landtree** [reply]
![image](https://github.com/user-attachments/assets/1173fd72-0f5d-49fc-b758-9523f2753d4d)

**Reply by Paciente8159** [reply]
Now we are getting somewhere. The code is were it was suppose to be.
But it's not running.

Just one last test please. Since you have the STLink in hand and already know how to re-flash the bootloader one last test please.
Modify the STM32 ini like I showed here. Compile.
Then use the STM32 tool to open the firmware.bin and flash the MCU directly to the default 0x08000000 address.
This will just to put the bootloader aside and check if it's a configuration issue (other then the bootloader).

See if it communicates.

This will overwrite your bootloader. You will have to re-flash it again after this test. 

> 
> ![image](https://private-user-images.githubusercontent.com/12116536/357973161-56d819f9-b7d3-453f-bbc3-147d63079fef.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM2NzA4ODIsIm5iZiI6MTcyMzY3MDU4MiwicGF0aCI6Ii8xMjExNjUzNi8zNTc5NzMxNjEtNTZkODE5ZjktYjdkMy00NTNmLWJiYzMtMTQ3ZDYzMDc5ZmVmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODE0VDIxMjMwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgxZTNiYmQ4NmQ3NDhkYzkyY2QyZjhiZjcwNjllOWIwZTIxMWJlZWM3M2JkOWJhM2FlZmVkZTEyOGQxMGI5ZWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.USbO0fCFJgA-lCxffVBxZ_lssxqxL3w-YPfcsl7qV5k)

**Reply by landtree** [reply]
That image is not loading, but I can do it asap!

**Reply by Paciente8159** [reply]
like this

```
[env:STM32F4-SKR-Pro-V1_2]
extends = common_stm32
board = black_f407zg
; board_build.flash_offset = 0x8000
board_build.f_cpu = 168000000L
; board_upload.offset_address = 0x08008000
board_upload.maximum_ram_size = 196608
board_upload.maximum_size = 1048576
build_flags = ${common_stm32.build_flags} -D BOARD=BOARD_SKR_PRO_V1_2 -D HSE_VALUE=8000000 -D RAM_ONLY_SETTINGS
```

**Reply by landtree** [reply]
No USB detection. 

![image](https://github.com/user-attachments/assets/d8111226-dfb4-42f8-911b-d73c8ff4ed88)
 &
 
![image](https://github.com/user-attachments/assets/53912838-83d3-451c-9a68-2fe0a0feef05)

**Reply by Paciente8159** [reply]
I can't thank you enough for this.
Then this only resumes to some sort of configuration (most probably clocks) issue that I will have to crack my head around.

I deeply appreciate the effort with all the tests.

**Reply by landtree** [reply]
No problem! Let me know if you want any other tests and I will do what I can.

**Reply by Paciente8159** [reply]
@landtree sorry for the silence. Been busy pushing v1.10 out.

One question. Have you tried GrblHAL?
Does it boot ok or does it have the same problem?

**Reply by landtree** [reply]
Not a problem! Don't sweat it. 

I double-confirmed and used the grblHal web builder, with the only mod being turning on auto square for Z and Y.

It can connect to USB.

**Reply by landtree** [reply]
@Paciente8159 Just as a sanity check to ensure I am not doing something wrong during the build. I downloaded the 1.10 release build and it still doesn't connect or see it on my computer. 

Thanks again for all the work you put into this!!

**Reply by Paciente8159** [reply]
Thanks for the tests.
You are building it without any config files correct ?from the web builder? Just the firmware as it is?

I will try to do a build using a custom linkerscript and upload it to my MkS board.
I'll then readjust it for your board and give it a go.

**Reply by landtree** [reply]
In my most recent test, I downloaded the build from the[ release page.](https://github.com/Paciente8159/uCNC/releases/download/v1.10.0/STM32F4-SKR-Pro-V1_2.zip) 

I have just tested by building on my end:

- Clean repo clone of 1.10, No config files, just the raw firmware.
- Clean repo clone of 1.10, Basic web builder configs (no changes)
- All uploaded via SD Card

All three resulted in the same: no blinking and no USB.

**Reply by Paciente8159** [reply]
I've seen some stuff over at grblhal github about the skr pro board with a 25MHz crystal and not an 8MHz as stated in the schematic.

Can you look at your board and check the crystal value please?

**Reply by Paciente8159** [reply]
Or you can simply try and upload the firmware by changing the `HSE_VALUE` to 25000000 in the stm32.ini file on the skr pro board settings

**Reply by landtree** [reply]
Okay, Looking at my board and the schematic on the [SKR Repo](https://github.com/bigtreetech/BIGTREETECH-SKR-PRO-V1.1/blob/master/SKR-PRO-V1.2/Schematic/SKR-PRO-V1.2.PDF).
![image](https://github.com/user-attachments/assets/b3711ff6-adec-4f92-8471-7417697400cb)


My Y3 is labeled as "YC8.0" and Y1 is empty.

**Reply by Paciente8159** [reply]
Check [this](https://github.com/grblHAL/STM32F4xx/issues/46#issuecomment-991801661)

**Reply by landtree** [reply]
Okay, I added this -D HSE_VALUE=25000000 and no change. 
Then I added "-D has_bootloader" and nothing.  Is it worth adding all the flags? I have on and off time tomorrow for testing.

![image](https://github.com/user-attachments/assets/a92a4adc-ebd7-4987-b415-f5057fa6cbf7)

**Reply by Paciente8159** [reply]
No.those other flags are specific to the grblhal.
I'll try to build a version using a makefile and send you the bin for you to try.

**Reply by Paciente8159** [reply]
Try this please.
[firmware.zip](https://github.com/user-attachments/files/16694781/firmware.zip)

**Reply by Paciente8159** [reply]
Also after trying that firmware I sent before, give a try at this branch. 🤞 
https://github.com/Paciente8159/uCNC/pull/738

**Reply by landtree** [reply]
@Paciente8159 That firmware file works!! We have blinking activity LED, I can connect over USB and was able to send G0 X50 and get movement. (That is currently the only axis that matches default setup).

I will try that branch right now and let you know!

**Reply by Paciente8159** [reply]
> @Paciente8159 That firmware file works!! We have blinking activity LED, I can connect over USB and was able to send G0 X50 and get movement. (That is currently the only axis that matches default setup).
> 
> I will try that branch right now and let you know!

Great news.
I was trying to run the firmware on my board using a manually compiled version (no Arduino no PlatformIO just a makefile), were I got full control of the code that is being compiled.
I was not being able to run anything either. After running an Excel spreadsheet from STM to calculate the configuration clocks I realized I had a mistake in the APB2 clocks division.

My board should have not worked in the first place. My guess why it was running OK is that the variant I used as a base board on my configuration sets the correct division for APB2 and on your board variant it didn't.

Also a configuration for flash access latency was adjusted but I believe the main problem was the APB2 division factor.

**Reply by landtree** [reply]
@Paciente8159  Okay, I compiled on that branch. (Stock no mods or configs).

No LED or USB. At least there is some momentum going!

**Comment by Paciente8159** [comment]
@landtree can you please try and change the stm32.ini file board variant for the skr pro from

`board = black_f407zg`

To 

`board = black_f407vg`

**Reply by landtree** [reply]
@Paciente8159 Adjusted and no LED or USB

**Reply by Paciente8159** [reply]
In that case there is only one test left to do. It's to add my custom linker script template.
I'll first try it on my board and I'll let you know.

**Reply by Paciente8159** [reply]
@landtree please try to download #738 again and upload the firmware.
I added a custom linker script and a complete new startup code.

My only question if this works is what implications this might have if one module uses Arduino libraries that depend on hardware.
But since most or all modules are Arduino independent the risk of it is minimal.

**Reply by landtree** [reply]
That did the trick. I was able to build from a clean download and can connect over USB and have the blinking LED.

No configs, just the base firmware.

**Comment by Paciente8159** [comment]
Awesome.
Tomorrow I will try to fine tune that configuration.
I may need for you to test it on your board if you don't mind.

Huge thank you for all your testing and patience.

**Reply by landtree** [reply]
No problem! I have availability again; there are just a couple of meetings sprinkled throughout the day.

**Reply by Paciente8159** [reply]
I've updated the linker script and tweaked a bit the vector table on #738.
Test on by board with success. I'm just waiting for the test on your hand to merge #738.
I've also checked the vector table IRQ names and this should work just fine with the Arduino framework.
Let me know how it goes.

**NOTE:**
I still don't have an explanation for why this is doing what is doing. Apart from the initial problem of the platformio configuration board parameters changed names between updates and the APB2 incorrect configuration, the code should simply work as it does on my board.
I guess the key might be on the way the bootloader works and configures the board initially.

**Reply by landtree** [reply]
Hey sorry it took so long. Day got crazy.

I downloaded #738   and used no config or adjustments. No response from board or USB.

**Reply by Paciente8159** [reply]
Uau I was not expecting that. Ok I'll revert then.

**Reply by Paciente8159** [reply]
I reverted the changes. Can you retry now?

**Reply by landtree** [reply]
This build works. USB & Activity LED.

**Reply by Paciente8159** [reply]
That is very intriguing.
Something in the linkerscript file then I assume.
I'll keep this version and investigate.

Thanks for all the tests and help.

**Reply by landtree** [reply]
No problem! I will try using a config file in a little bit to see if anything explodes! I will let you know if I run into any problems.

**Reply by landtree** [reply]
@Paciente8159 I just did a quick test with the config files. Other than some tweaks to have the correct step direction & speed, it seems like a functional build!

Thanks again!
